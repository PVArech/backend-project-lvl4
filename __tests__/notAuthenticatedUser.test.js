import _ from 'lodash';
import getApp from '../server/index.js';
import { getTestData, prepareData } from './helpers/index.js';

describe('tests not Authenticated', () => {
  let app;
  let knex;
  const testData = getTestData();
  const {
    verbs,
    routes,
    routesNew,
    routesEdit,
    routesId,
  } = testData;
  const id = 1;
  const status = {
    GET: 302,
    POST: 302,
    PATCH: 404,
    DELETE: 404,
  };

  beforeAll(async () => {
    app = await getApp();
    knex = app.objection.knex;
    await knex.migrate.latest();
    await prepareData(app);
  });

  describe.each(_.difference(verbs, ['PATCH', 'DELETE']))('test %s /route', (verb) => {
    it.each(routes)(`${verb} /%s`, async (route) => {
      const response = await app.inject({
        method: verb,
        url: app.reverse(route),
        payload: verb === 'POST' ? { data: testData[route].new } : '',
      });
      const expected = (route === 'users' && verb === 'GET') ? 200 : status[verb];
      expect(response.statusCode).toBe(expected);
    });
  });

  describe('test GET /route/new', () => {
    it.each(routesNew)('GET /%s', async (route) => {
      const response = await app.inject({
        method: 'GET',
        url: app.reverse(route),
      });
      const expected = route === 'newUser' ? 200 : 302;
      expect(response.statusCode).toBe(expected);

      const tree = response.body;
      expect(tree).toMatchSnapshot();
    });
  });

  describe('test GET /route/:id/edit', () => {
    it.each(routesEdit)('GET /%s', async (route) => {
      const response = await app.inject({
        method: 'GET',
        url: app.reverse(route, { id }),
      });
      expect(response.statusCode).toBe(302);
    });
  });

  describe.each(_.difference(verbs, ['POST']))('test %s /route/:id', (verb) => {
    it.each(routesId)(`test ${verb} /%s=${id}`, async (route) => {
      const response = await app.inject({
        method: verb,
        url: app.reverse(route, { id }),
      });
      const expected = (route !== 'task' && verb === 'GET') ? 404 : 302;
      expect(response.statusCode).toBe(expected);
    });
  });

  afterAll(async () => {
    await knex.migrate.rollback();
    app.close();
  });
});
