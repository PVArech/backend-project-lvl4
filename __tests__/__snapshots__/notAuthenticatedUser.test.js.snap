// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests not Authenticated test GET /route/new GET /newLabel 1`] = `""`;

exports[`tests not Authenticated test GET /route/new GET /newStatus 1`] = `""`;

exports[`tests not Authenticated test GET /route/new GET /newTask 1`] = `""`;

exports[`tests not Authenticated test GET /route/new GET /newUser 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><title>Hexlet Fastify Boilerplate</title><script src=\\"/assets/main.js\\"></script><link href=\\"/assets/main.css\\" rel=\\"stylesheet\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\"></head><body class=\\"d-flex flex-column min-vh-100\\"><nav class=\\"navbar navbar-expand-lg navbar-light bg-light mb-3\\"><div class=\\"container-fluid\\"><a class=\\"navbar-brand\\" href=\\"/\\">Менеджер задач</a><button class=\\"navbar-toggler\\" data-bs-toggle=\\"collapse\\" data-bs-target=\\"#navbarToggleExternalContent\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarToggleExternalContent\\"><div class=\\"container-fluid\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/users\\">Пользователи</a></li></ul></div><ul class=\\"navbar-nav\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/session/new\\">Вход</a></li><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/users/new\\">Регистрация</a></li></ul></div></div></nav><div class=\\"container wrapper flex-grow-1\\"><h1 class=\\"my-4\\">Регистрация</h1><form action=\\"/users\\" method=\\"post\\"><div class=\\"mb-3\\"><label for=\\"data_firstName\\">Имя</label><input class=\\"form-control\\" id=\\"data_firstName\\" name=\\"data[firstName]\\" value=\\"\\" type=\\"text\\"/></div><div class=\\"mb-3\\"><label for=\\"data_lastName\\">Фамилия</label><input class=\\"form-control\\" id=\\"data_lastName\\" name=\\"data[lastName]\\" value=\\"\\" type=\\"text\\"/></div><div class=\\"mb-3\\"><label for=\\"data_email\\">Email</label><input class=\\"form-control\\" id=\\"data_email\\" name=\\"data[email]\\" value=\\"\\" type=\\"text\\"/></div><div class=\\"mb-3\\"><label for=\\"data_password\\">Пароль</label><input class=\\"form-control\\" id=\\"data_password\\" name=\\"data[password]\\" value=\\"\\" type=\\"password\\"/></div><input class=\\"btn btn-primary\\" type=\\"submit\\" value=\\"Сохранить\\"/></form></div><footer><div class=\\"container my-5 pt-4 border-top\\"><a target=\\"_blank\\" href=\\"https://ru.hexlet.io\\">Hexlet</a></div></footer></body></html>"`;
